pyo3-derive-backend/src/pyfunction.rs

```diff
@@ -94,97 +83,82 @@ impl PyFunctionAttr {
     }
 
     fn add_work(&mut self, item: &NestedMeta, path: &Path) -> syn::Result<()> {
-        // self.arguments in form somename
-        if self.has_kwargs {
+        if self.has_kw || self.has_kwargs {
             return Err(syn::Error::new_spanned(
                 item,
-                "syntax error, keyword self.arguments is defined",
+                "Positional argument or varargs(*) is not allowed after keyword arguments",
             ));
         }
-        if self.has_kw {
+        if self.has_varargs {
             return Err(syn::Error::new_spanned(
                 item,
-                "syntax error, argument is not allowed after keyword argument",
+                "Positional argument or varargs(*) is not allowed after *",
             ));
         }
         self.arguments.push(Argument::Arg(path.clone(), None));
         Ok(())
     }
 
+    fn vararg_is_ok(&self, item: &NestedMeta) -> syn::Result<()> {
+        if self.has_kwargs || self.has_varargs {
+            return Err(syn::Error::new_spanned(
+                item,
+                "* is not allowed after varargs(*) or kwargs(**)",
+            ));
+        }
+        Ok(())
+    }
+
+    fn kw_arg_is_ok(&self, item: &NestedMeta) -> syn::Result<()> {
+        if self.has_kwargs {
+            return Err(syn::Error::new_spanned(
+                item,
+                "Keyword argument or kwargs(**) is not allowed after kwargs(**)",
+            ));
+        }
+        Ok(())
+    }
+
+    fn add_nv_common(
```

Comment:
This function is just meant for refactoring.
Seprated the common parts from `add_name_value`.